File: Leetcode_with_Cpp/78subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        if (nums.empty()) return {{}};
        int n = nums.back();
        nums.pop_back();
        vector<vector<int>> res = subsets(nums);
        int sz = res.size();
        for (int i = 0; i < sz; i++) {
            res.push_back(res[i]);
            res.back().push_back(n);
        }
        return res;
    }
};

class Solution {
private:
    vector<vector<int>> res;
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> track;
        backtrack(nums, 0, track);
        return res;
    }
    void backtrack(vector<int>& nums, int start, vector<int>& track) {
        res.push_back(track);
        for (int i = start; i < nums.size(); i++) {
            track.push_back(nums[i]);
            backtrack(nums, i + 1, track);
            track.pop_back();
        }
    }
};

File: py/78subsets.py
class Solution(object):
    def subsets(self, nums):
        res = [[]]
        for i in range(len(nums)-1, -1, -1):
            for subres in res[:]:
                res.append(subres + [nums[i]])
        return res

# 从后遍历，遇到一个数就把所有子集加上该数组成新的子集
# 在循环中res[:]不更新，而res不断有新元素push进去